### PR TITLE
nixos/gatus: init module

### PR DESCRIPTION
**Pull request by pizzapim**
## Description of changes

In https://github.com/NixOS/nixpkgs/pull/269477, the Gatus package was added to nixpkgs. I created a NixOS module to run Gatus as a Systemd service. I have been running this myself for a week with no problems. You can see on my personal Git server how this can be used: [here](https://git.kun.is/home/nixos-servers/src/commit/01d54f8c2e0051cb688cb70fd0a953505fd4cea3/nix/modules/monitoring/gatus-endpoints.nix) I define Gatus endpoints which are services I monitor, and [here](https://git.kun.is/home/nixos-servers/src/commit/01d54f8c2e0051cb688cb70fd0a953505fd4cea3/nix/modules/monitoring/default.nix#L56) I configure the service itself.

This is my first contribution to nixpkgs so any comments are greatly appreciated!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NyCodeGHG** [comment]
Hi! Welcome and thank you for your first contribution :)

Please squash your commits into a single commit and use a proper commit message. 
See the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/nixos/README.md#commit-conventions)

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:20]: you can remove a level of indent here
```suggestion
  options.services.gatus = {
```

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:37]: not sure if this is needed when settings exist?

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:106]: you might want to move this as the default value for configFile, so you can remove a lot of checks and the weird indent
(if we decide to keep configFile, i'm not sure about that)

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:122]:
```suggestion
          EnvironmentFile = lib.optional (cfg.environmentFile != null) cfg.environmentFile;
```

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:137]:
```suggestion
      users.users.gatus = mkIf (cfg.user == "gatus") {
        isSystemUser = true;
        group = cfg.group;
        home = "/var/lib/${config.systemd.services.gatus.serviceConfig.StateDirectory}";
      };
```

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:141]:
```suggestion
      users.groups.gatus = mkIf (cfg.group == "gatus") { };
```

File: nixos/modules/services/monitoring/gatus.nix
Comment (NyCodeGHG) [inline nixos/modules/services/monitoring/gatus.nix:145]:
```suggestion
      networking.firewall.allowedTCPPorts = lib.optionals cfg.openFirewall [ cfg.settings.web.port ];
```

**Review by NyCodeGHG — CHANGES_REQUESTED** [review]

**Comment by NyCodeGHG** [comment]
I noticed that gatus also supports postgresql as a storage backend, an option for that would be nice, but could also be done in a future PR (and also by someone else, if you don't plan to use it).

File: nixos/modules/services/monitoring/gatus.nix
Comment (drupol) [inline nixos/modules/services/monitoring/gatus.nix:23]: I think you can replace it with: 

```suggestion
      package = mkPackageOption pkgs "gatus" { };
```

**Comment by nu-nu-ko** [comment]
none of this is killer so felt iffy for review buuut
`mdDoc` is pointless and will be removed tree wide eventually anyway.
restart always may be better as unless stopped?
and personal preference not everyone seems to agree with, another `inherit` for `(lib.types) foo bar` would be nicer than the using `with` at all
also why not add yourself as a maintainer?

File: nixos/modules/services/monitoring/gatus.nix
Comment (drupol) [inline nixos/modules/services/monitoring/gatus.nix:136]: Would it be possible to use a dynamic user instead?

Comment (pizzapim) [reply nixos/modules/services/monitoring/gatus.nix:136]: I didn't know of this feature, I think this will work.

**Comment by pizzapim** [comment]
> I noticed that gatus also supports postgresql as a storage backend, an option for that would be nice, but could also be done in a future PR (and also by someone else, if you don't plan to use it).

Indeed, for my use case SQLite is more than enough, so I'd say we can leave this for a future PR of (probably) someone else :) Does seem like a good idea though.

File: nixos/modules/services/monitoring/gatus.nix
Comment (drupol) [reply nixos/modules/services/monitoring/gatus.nix]: Better and better at each iteration !

Comment (pizzapim) [reply nixos/modules/services/monitoring/gatus.nix]: Doesn't this shift the burden of user/group creation to the user though? If the user wants persistent state, they need to have a user/group with predictable UID/GID. With this change, now the user is responsible for arranging this. But maybe this is a non-argument?

Comment (pizzapim) [reply nixos/modules/services/monitoring/gatus.nix]: Now that I think about it, I think the original setup didn't have predictable UID/GID anyways so this might be fine :)

**Comment by pizzapim** [comment]
Thanks for all the reviews! Unless there are further comments I think this might be ready for a merge?

**Comment by NyCodeGHG** [comment]
i have one small thing, would be nice if you could create an extra commit for creating your maintainer in maintainer-list.nix
with the commit message `maintainers: add pizzapim`, other than that this should be ready to merge :)

**Comment by pizzapim** [comment]
Done and done!

**Review by NyCodeGHG — APPROVED** [review]
lgtm

**Comment by patcher-ms** [comment]
Hello @pizzapim thanks for the work you put in this PR. What’s left to do before it gets merged?

**Comment by NyCodeGHG** [comment]
The merge conflict needs to be resolved by rebasing on master, then it should be ready to merge

File: nixos/modules/services/monitoring/gatus.nix
Comment (drupol) [inline nixos/modules/services/monitoring/gatus.nix:1]: Could you please pass this new file through `nixfmt-rfc-style` ?

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/modules/services/monitoring/gatus.nix
Comment (pizzapim) [reply nixos/modules/services/monitoring/gatus.nix]: Done!

**Comment by patcher-ms** [comment]
@drupol, sorry for the ping. The changes you requested have been made.
Could you please have another look. Is there anything else blocking this from being merged?

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (drupol) [inline nixos/doc/manual/release-notes/rl-2405.section.md:98]: Isn't this supposed to `nixos/doc/manual/release-notes/rl-2411.section.md` ?

Comment (pizzapim) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Sorry for the delay, I have moved the release notes to 24.11

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
It would be nice to have some tests. Just a basic NixosTest that would check if the service is running. If you could do this for the next PR it would be nice, I won't block further this PR.

File: nixos/tests/all-tests.nix
Comment (drupol) [inline nixos/tests/all-tests.nix:368]: Please use `runTest`, it's easier.

File: nixos/tests/gatus.nix
Comment (drupol) [inline nixos/tests/gatus.nix:1]: This will be removed when using `runTest`. Have a look at some examples there.

File: nixos/tests/gatus.nix
Comment (drupol) [inline nixos/tests/gatus.nix:28]: Remove the comment.

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/tests/gatus.nix
Comment (pizzapim) [reply nixos/tests/gatus.nix:28]: The comment is to instruct treesitter about the language inside the comment, but I see this is not compatible with formatting :(

**Review by drupol — CHANGES_REQUESTED** [review]
Almost perfect, now you can add those `nixosTest` to the `gatus` package. Have a look at `honk` that does that.

**Review by drupol — CHANGES_REQUESTED** [review]
The commit about `pkgs/by-name/ga/gatus/package.nix` needs to be separated and with its own commit log message, e.g.,  `gatus: add nixosTests`

**Comment by pizzapim** [comment]
I think the test is broken for Darwin for some reason?

**Comment by pizzapim** [comment]
@drupol could this be merged now? I read on [the forums](https://discourse.nixos.org/t/why-is-it-taking-so-long-to-handle-pr-adding-package/52734/5) that the Darwin tests are known to be flakey.

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks for the reminder, LGTM now.

**Pull request merged**
